### ISSUE TITLE
Error in PostgreSQL connection: Error { kind: Closed, cause: None }

### ISSUE THREAD

**Issue by Ezard** [prisma/prisma#6791]
Hi Prisma Team! My Prisma Client just crashed. This is the report:
## Versions

| Name            | Version            |
|-----------------|--------------------|
| Node            | v14.16.0           | 
| OS              | debian-openssl-1.1.x|
| Prisma Client   | 2.21.2             |
| Query Engine    | query-engine e421996c87d5f3c8f7eeadd502d4ad402c89464d|
| Database        | undefined|

## Query
```
{"X":true}
```

## Logs
```
tdout  Unknown error
prisma:engine stdout  Unknown error
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Unknown error
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Unknown error
prisma:engine stdout  Unknown error
prisma:engine stdout  Error in PostgreSQL connection: Error { kind: Closed, cause: None }
prisma:engine stdout  Unknown error
prisma:engine stdout  Unknown error
prisma:engine {
  error: Error: Error in connector: Server terminated the connection.
      at /app/node_modules/@prisma/client/runtime/index.js:27438:19
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
}
prisma:engine stdout  Fetched a connection from the pool
prisma:engine stdout  Error in PostgreSQL connection: Error { kind: Io, cause: Some(Os { code: 104, kind: ConnectionReset, message: "Connection reset by peer" }) }
prisma:engine stdout  Unknown error
prisma:engine stdout  Error in PostgreSQL connection: Error { kind: Closed, cause: None }
prisma:engine {
  error: Error: Error in connector: Server terminated the connection.
      at /app/node_modules/@prisma/client/runtime/index.js:27438:19
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
}
prisma:engine {
  error: Error: Error in connector: Server terminated the connection.
      at /app/node_modules/@prisma/client/runtime/index.js:27438:19
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
}
```

## Additional details:
- the application code is running on Google Kubernetes Engine
- database is a private Postgres Cloud SQL instance, which the application code is connected to directly (i.e. not via the Cloud SQL Proxy) over the VPC via a private IP address
- SSL is used
- the issue was not caused by connection starvation; around the time of the issue the maximum number of connections open on the DB was 14 (including the connections reserved by Cloud SQL itself), and the maximum number of connections that this database instance supports is 25
- at the time of the issue, the DB instance's CPU usage was at ~4%, and its memory usage was at ~60%
- the issue occurred only once, and then Prisma continued to work as expected within a few milliseconds

More than happy to pull more data from the monitoring dashboards if needed

Similar issues to this have been raised a few times before (all of which are using GCP with Postgres), but most of these have been closed. I intend to make use of the retry middleware suggested [here](https://github.com/prisma/prisma/issues/6329#issuecomment-816954832) since this issue only occurs once or twice a month for me, but I figured I may as well report it in case there's info the Prisma team can make use of :)

**Comment by PranavSathy** [comment]
@Ezard Out of curiosity, did you enable "Cloud SQL Insights" on your Cloud SQL instance? We had the -exact- same issue when Cloud SQL Insights was enabled. IF you look into your Cloud Monitoring/Logging dashboard and filter for log statements from the cloud SQL instance, it's likely you will notice a `segmentation fault`. My current hypothesis is that the Cloud SQL Insights pgSQL plugin has some faulty logic causing in this segmentation fault.

If you do have it enabled, the fix for us was simply disabling it. You can leverage tools like NewRelic / AppDynamics / DataDog to monitor your SQL performance.

**Comment by Ezard** [comment]
@PranavSathy ahhh - we have Query Insights enabled, yes
I just checked the logs and found the `segmentation fault` message ~1 second before the Prisma error appeared
Weird that the seg fault isn't reported as an error in their logging, but hey ho

I've found a few open issues for this in Google's bugtracker, for reference:
https://issuetracker.google.com/issues/184065841
https://issuetracker.google.com/issues/184283279
https://issuetracker.google.com/issues/179716520

For the Prisma team: this issue doesn't seem to have anything to do with Prisma itself (which I'm sure you're glad to hear!); I'm unsure if you want to close it or leave it open for other people having the same issue (until Google has fixed their product, anyway), so I'll leave that up to you

**Comment by janpio** [comment]
I wonder if we could do something here to surface this problem to Prisma and Google Cloud SQL users... The error message alone does not seem to be enough, as it is a normal error message when the server goes away.

**Comment by pantharshit00** [comment]
Looks like this is going to get patched on 28th May
![image](https://user-images.githubusercontent.com/22195362/119279515-d9582880-bc49-11eb-994b-2fc5daa48a7c.png)

**Comment by Ezard** [comment]
I haven't seen this crash for several months now, so I'll close this issue since it's no longer relevant
